---
title: "01_CTCF_WASH_RNASE"
author: "JR"
date: "8/11/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{bash}

#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=CTCF_macs
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=john.rinn@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=100gb
#SBATCH --time=6:00:00
#SBATCH --output=CTCF_chip_macs.out
#SBATCH --error=CTCF_chip_macs.err
date; hostname

# going to nf_core chipseq outputs
BASE_PATH=/scratch/Shares/rinn/JR/rchip/analysis/03_diff_peaks/data/BAM_files/ctcf

# calling out macs3 to avtivate an env
source activate macs3

# callpeak to call peaks
# -f means bam file paired end 
# -t is treatment group / -c is control group
# -g is the genome "mm"
# -n is the name of the output file
# -- broad peak (narrow peak is the default)
# --broad-cutoff is distance between peaks to be merged
# -B probably makes a beggraph file for the outputs

macs3 callpeak -f BAMPE -t \
${BASE_PATH}/3_ctcf_wash_rnase_rep_1.mLb.clN.sorted.bam \
${BASE_PATH}/3_ctcf_wash_rnase_rep_2.mLb.clN.sorted.bam \
${BASE_PATH}/3_ctcf_wash_rnase_rep_3.mLb.clN.sorted.bam \
-c \
${BASE_PATH}/4_input_rep_1.mLb.clN.sorted.bam \
${BASE_PATH}/4_input_rep_2.mLb.clN.sorted.bam \
${BASE_PATH}/4_input_rep_3.mLb.clN.sorted.bam \
${BASE_PATH}/4_input_rep_4.mLb.clN.sorted.bam \
${BASE_PATH}/4_input_rep_5.mLb.clN.sorted.bam \
-g hs -n CTCF_chipseq_combined  -q 0.01 --outdir ../results/ctcf_wash_rnase
date

```

