---
title: "01_CTCF_WASH_RNASE"
author: "JR"
date: "8/11/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
library(Rsubread)
library(rtracklayer)
library(DESeq2)
```


# (1) run macs in bin dir
```{bash}

#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=CTCF_macs
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=john.rinn@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=100gb
#SBATCH --time=6:00:00
#SBATCH --output=CTCF_chip_macs.out
#SBATCH --error=CTCF_chip_macs.err
date; hostname

# going to nf_core chipseq outputs
BASE_PATH=/scratch/Shares/rinn/JR/rchip/analysis/03_diff_peaks/data/BAM_files/ctcf

# calling out macs3 to avtivate an env
source activate macs3

# callpeak to call peaks
# -f means bam file paired end 
# -t is treatment group / -c is control group
# -g is the genome "mm"
# -n is the name of the output file
# -- broad peak (narrow peak is the default)
# --broad-cutoff is distance between peaks to be merged
# -B probably makes a bedgraph file for the outputs

macs3 callpeak -f BAMPE -t \
${BASE_PATH}/1_ctcf_chip_rep_1.mLb.clN.sorted.bam \
${BASE_PATH}/1_ctcf_chip_rep_2.mLb.clN.sorted.bam \
${BASE_PATH}/1_ctcf_chip_rep_3.R1.mLb.clN.sorted.bam \
${BASE_PATH}/1_ctcf_chip_rep_4.R1.mLb.clN.sorted.bam \
${BASE_PATH}/1_ctcf_chip_rep_5.R1.mLb.clN.sorted.bam \
-c \
${BASE_PATH}/4_input_rep_1.mLb.clN.sorted.bam \
${BASE_PATH}/4_input_rep_2.mLb.clN.sorted.bam \
${BASE_PATH}/4_input_rep_3.R1.mLb.clN.sorted.bam \
${BASE_PATH}/4_input_rep_4.R1.mLb.clN.sorted.bam \
${BASE_PATH}/4_input_rep_5.R1.mLb.clN.sorted.bam \
-g hs -n CTCF_chipseq_combined  -q 0.01 --outdir ../results/ctcf_chip_combined_5_reps
date

```


# (2) Create simple annotation file (.saf) for input into FeatureCounts
```{r loading in the consensus peak file (- RNase ) }

# loading in the narrow peak consensus peaks between two CTCF ChIPseq replicates (June 2022)
# Two inputs compared to two replicates and run on narrow peak instead of NF_CORE output broadPeak

ctcf_peaks <- import("results/ctcf_chip_combined_5_reps/CTCF_chipseq_combined_peaks.narrowPeak")

# The input to FeatureCounts is .saf (simple annotation file) which will quantify the reads in each peak
# we need a SAF file needs that requires these colnames:
# GeneID		Chr	Start	End	Strand
# creating SAF file below

ctcf_saf <- ctcf_peaks %>%
  as.data.frame() %>%
  # making sure to only grab normal chr not contigs (e.g., GL000009.2)
  filter(grepl("chr", seqnames)) %>%
  # renaming cols to fit format of .saf
  dplyr::rename(GeneID = name,
                Chr = seqnames,
                Start = start,
                End = end,
                Strand = strand) %>%
    dplyr::select(GeneID, Chr, Start, End, Strand)

# Writting out .saf
write.table(ctcf_saf, "results/ctcf_chip_combined_5_reps/ctcf_peaks_combined.saf", sep = "\t", quote = F, row.names = F)

```

# start point (saf file)
```{r}
read.table()

```


# (3) Run feature counts on all bam files +/- rnase
# This will provide quntification of reads in each peak in consensus peaks
# For each sample -- these values can be used as input for deseq below
```{r feature counts across all bam files (+/- RNase)}

# create an object of all bam files (*** including inputs ***)
ctcf_bams <- list.files("data/BAM_files/ctcf", full.names = T)

# Running feature counts on the bam files from ENCODE that are single ended (-RNase)
# default is single ended but added isPairedEnd parameter to False

ctcf_feature_counts <- featureCounts(files = ctcf_bams,
                    annot.ext = "results/ctcf_chip_combined_5_reps/ctcf_peaks_combined.saf",
                    isPairedEnd = TRUE,
                    nthreads = 6)


# saving
write_rds(ctcf_feature_counts, "results/ctcf_chip_combined_5_reps/CTCF_feature_counts_output.rds")


```

# start point (with feature counts)
```{r}

ctcf_feature_counts <- read_rds("results/ctcf_chip_combined_5_reps/CTCF_feature_counts_output.rds")

```


# creating sample_df
```{r}

counts_matrix <- ctcf_feature_counts$counts



sample_df <- data.frame(filename = ctcf_feature_counts[["targets"]]) %>%
  mutate(sample_name = filename)

  sample_df$sample_name = gsub(".bam|.mLb.clN.sorted.bam|R1.mLb.clN.sorted.bam", "", sample_df$sample_name) 
  
  sample_df <- sample_df %>%
    mutate(condition = ifelse(grepl("wash", sample_name), "rnase_wash", ifelse(grepl("chip", sample_name), "ctcf_chip", ifelse(grepl("input", sample_name), "input", "rnase"))))
  
  samples_rnase <- sample_df %>%
    filter(condition %in% c("rnase", "ctcf_chip"))
  
  rnase_count_matrix <- counts_matrix[, samples_rnase$filename]
  
  
  # factoring for deseq
  
  samples_rnase <- samples_rnase %>%
    mutate(condition = factor(condition, levels = c("ctcf_chip", "rnase")))
  
  

  # Deseq
  dds_rnase <- DESeqDataSetFromMatrix(countData = rnase_count_matrix,
                              colData = samples_rnase,
                              # diff on snyder IPs (2) vs rinn IPs (2)
                              design = ~ condition)
# Running Deseq
dds_rnase <- DESeq(dds_rnase)

# getting names
resultsNames(dds_rnase)

# filter out intercept results
rnase_res <- results(dds_rnase, name = "condition_rnase_vs_ctcf_chip", tidy = TRUE)
  


ggplot(rnase_res, aes(x = log2FoldChange, y = -log10(pvalue))) +
         geom_point()





```



```{r}




samples_chip <- sample_df %>%
    filter(condition %in% c("input", "ctcf_chip"))
  
  chip_count_matrix <- counts_matrix[, samples_rnase$filename]
  
  
  # factoring for deseq
  
  samples_chip <- samples_rnase %>%
    mutate(condition = factor(condition, levels = c("input", "ctcf_chip")))
  
  
  
  # chip 
dds_chip_input <- DESeqDataSetFromMatrix(countData = chip_count_matrix,
                              colData = samples_chip,
                              # diff on snyder IPs (2) vs rinn IPs (2)
                              design = ~ condition)

dds_chip_input <- DESeq(dds_chip_input)

# getting names
resultsNames(dds_chip_input)

# filter out intercept results
chip_input_res <- results(dds_chip_input, name = "condition_ctcf_chip_vs_input", tidy = TRUE)
  

  
```


# Rnase 
```{r}

samples_rnase_input <- sample_df %>%
    filter(condition %in% c("input", "rnase"))
  
  rnase_input_count_matrix <- counts_matrix[, samples_rnase_input$filename]
  
  
  # factoring for deseq
  
  samples_rnase_input <- samples_rnase_input %>%
    mutate(condition = factor(condition, levels = c("input", "rnase")))
  
  
  
  # chip 
dds_rnase_input <- DESeqDataSetFromMatrix(countData = rnase_input_count_matrix,
                              colData = samples_rnase_input,
                              # diff on snyder IPs (2) vs rinn IPs (2)
                              design = ~ condition)

dds_rnase_input <- DESeq(dds_rnase_input)

# getting names
resultsNames(dds_rnase_input)

# filter out intercept results
rnase_input_res <- results(dds_rnase_input, name = "condition_rnase_vs_input", tidy = TRUE)




# 
# colnames(chip_input_res)[1] <- "GeneID"
# 
# chip_input_res <- left_join(chip_input_res, ctcf_saf)





```






```{r }


chip_rnase_data <- rnase_res %>%
  dplyr::select(row, log2FoldChange, pvalue) %>%
  dplyr::rename(l2fc_rnase_vs_chip = log2FoldChange,
               pvalue_rnase_vs_chip = pvalue) %>%
  left_join(chip_input_res %>%
              dplyr::select(row, log2FoldChange, padj) %>%
              dplyr::rename(l2fc_chip_vs_input = log2FoldChange,
                            padj_chip_vs_input = padj)) %>%
  left_join(rnase_input_res %>%
              dplyr::select(row, log2FoldChange, padj) %>%
              dplyr::rename(l2fc_rnase_vs_input = log2FoldChange,
                            padj_rnase_vs_input = padj))
chip_rnase_data <- chip_rnase_data %>%
  mutate(sig = pvalue_rnase_vs_chip <= 0.05)

colnames(chip_rnase_data)[1] <- "GeneID"

chip_rnase_data <- left_join(chip_rnase_data, ctcf_saf)

# actually plotting
ggplot(chip_rnase_data, aes(x = l2fc_chip_vs_input, y = l2fc_rnase_vs_input, color = sig)) +
  geom_point() +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  geom_abline(slope = 1)







```







# (4) Deseq to determine peaks that are differntial between conditions
