---
title: "01_CTCF_WASH_RNASE"
author: "JR"
date: "8/11/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
library(Rsubread)
library(rtracklayer)
library(DESeq2)
```


# (1) run macs in bin dir
```{bash}

#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=CTCF_macs
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=john.rinn@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=100gb
#SBATCH --time=6:00:00
#SBATCH --output=CTCF_chip_macs.out
#SBATCH --error=CTCF_chip_macs.err
date; hostname

# going to nf_core chipseq outputs
BASE_PATH=/scratch/Shares/rinn/JR/rchip/analysis/03_diff_peaks/data/BAM_files/ctcf

# calling out macs3 to avtivate an env
source activate macs3

# callpeak to call peaks
# -f means bam file paired end 
# -t is treatment group / -c is control group
# -g is the genome "mm"
# -n is the name of the output file
# -- broad peak (narrow peak is the default)
# --broad-cutoff is distance between peaks to be merged
# -B probably makes a bedgraph file for the outputs

macs3 callpeak -f BAMPE -t \
${BASE_PATH}/1_ctcf_chip_rep_1.mLb.clN.sorted.bam \
${BASE_PATH}/1_ctcf_chip_rep_2.mLb.clN.sorted.bam \
${BASE_PATH}/1_ctcf_chip_rep_3.R1.mLb.clN.sorted.bam \
${BASE_PATH}/1_ctcf_chip_rep_4.R1.mLb.clN.sorted.bam \
${BASE_PATH}/1_ctcf_chip_rep_5.R1.mLb.clN.sorted.bam \
-c \
${BASE_PATH}/4_input_rep_1.mLb.clN.sorted.bam \
${BASE_PATH}/4_input_rep_2.mLb.clN.sorted.bam \
${BASE_PATH}/4_input_rep_3.R1.mLb.clN.sorted.bam \
${BASE_PATH}/4_input_rep_4.R1.mLb.clN.sorted.bam \
${BASE_PATH}/4_input_rep_5.R1.mLb.clN.sorted.bam \
-g hs -n CTCF_chipseq_combined  -q 0.01 --outdir ../results/ctcf_chip_combined_5_reps
date

```


# (2) Create simple annotation file (.saf) for input into FeatureCounts
```{r loading in the consensus peak file (- RNase ) }

# loading in the narrow peak consensus peaks between two CTCF ChIPseq replicates (June 2022)
# Two inputs compared to two replicates and run on narrow peak instead of NF_CORE output broadPeak

ctcf_peaks <- import("results/ctcf_chip_combined_5_reps/CTCF_chipseq_combined_peaks.narrowPeak")

# The input to FeatureCounts is .saf (simple annotation file) which will quantify the reads in each peak
# we need a SAF file needs that requires these colnames:
# GeneID		Chr	Start	End	Strand
# creating SAF file below

ctcf_saf <- ctcf_peaks %>%
  as.data.frame() %>%
  # making sure to only grab normal chr not contigs (e.g., GL000009.2)
  filter(grepl("chr", seqnames)) %>%
  # renaming cols to fit format of .saf
  dplyr::rename(GeneID = name,
                Chr = seqnames,
                Start = start,
                End = end,
                Strand = strand) %>%
    dplyr::select(GeneID, Chr, Start, End, Strand)

# Writting out .saf
write.table(ctcf_saf, "results/ctcf_chip_combined_5_reps/ctcf_peaks_combined.saf", sep = "\t", quote = F, row.names = F)

```

# (3) Run feature counts on all bam files +/- rnase
# This will provide quntification of reads in each peak in consensus peaks
# For each sample -- these values can be used as input for deseq below
```{r feature counts across all bam files (+/- RNase)}

# create an object of all bam files (*** including inputs ***)
ctcf_bams <- list.files("data/BAM_files/ctcf", full.names = T)

# Running feature counts on the bam files from ENCODE that are single ended (-RNase)
# default is single ended but added isPairedEnd parameter to False

ctcf_feature_counts <- featureCounts(files = ctcf_bams,
                    annot.ext = "results/ctcf_chip_combined_5_reps/ctcf_peaks_combined.saf",
                    isPairedEnd = TRUE,
                    nthreads = 6)


# saving
write_rds(ctcf_feature_counts, "results/ctcf_chip_combined_5_reps/CTCF_feature_counts_output.rds")

```
















# (4) Deseq to determine peaks that are differntial between conditions
```{r running deseq between our CTCF and encode CTCF two replicates each}

# creating a dataframe to be used in Deseq (DESeqDataSetFormMatrix)
sample_df <- data.frame(filename = colnames(ctcf_feature_counts)) %>%
  # Cleanup colnames from peak_counts
  mutate(sample_name = filename),
         sample_name = gsub(".bam|.R1.mLb.clN.sorted|.shearing", "", sample_name),
         sample_name = gsub("CTCF.10.20", "rinn.CTCF.rep1", sample_name),
         sample_name = gsub("CTCF.20.60", "rinn.CTCF.rep2", sample_name)) %>%
  separate(sample_name, into = c("experiment", "antibody", "replicate", "condition")) %>%
  replace_na(list(condition = "IP"))

# Snyder vs Rinn differential test of IP samples 
# In the future this will be RNase+ vs RNase-
# The input is consensus peaks derived from RNase - condition
sample_df1 <- sample_df %>%
  filter(experiment %in% c("snyder", "rinn"),
         condition == "IP") %>%
  mutate(experiment = factor(experiment, levels = c("rinn", "snyder")))

# indexing peak counts from feature counts to snyder and rinn samples
peak_counts1 <- peak_counts[,sample_df1$filename]

# setting up experimental design for + or - RNase (currently snyder vs rinn : experiment col)
dds1 <- DESeqDataSetFromMatrix(countData = peak_counts1,
                              colData = sample_df1,
                              # diff on snyder IPs (2) vs rinn IPs (2)
                              design = ~ experiment)
# Running Deseq
dds1 <- DESeq(dds1)

# getting names
resultsNames(dds1)

# filter out intercept results
res1 <- results(dds1, name = "experiment_snyder_vs_rinn", tidy = TRUE)

