---
title: "Differential peak calling"
author: "JR"
date: "6/29/2022"
output: git_hub 
editor_options: 
  chunk_output_type: console
mainfont:  Lucida Console
---


```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(Rsubread)
library(rtracklayer)
library(DESeq2)
```

# The first step in calling differential peaks is to have a consensus peak set called
# for the non-rnase treated samples.Since we want to be careful about peak boundaries, it's best to call peaks on the
# merge dataset.


# In order to call merged peaks, we'll run MACS3 with bin/call_merged_peak.sh

# (1) Running MACS3
# Here is an example .sh file

```{bash}
#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=CTCF_macs
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=john.rinn@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=100gb
#SBATCH --time=3:00:00
#SBATCH --output=CTCF_chip_macs.out
#SBATCH --error=CTCF_chip_macs.err
date; hostname

# going to nf_core chipseq outputs
BASE_PATH=/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary

# calling out macs3 to avtivate an env
source activate macs3

# callpeak to call peaks
# -f means bam file paired end 
# -t is treatment group / -c is control group
# -g is the genome "mm"
# -n is the name of the output file
# -- broad peak (narrow peak is the default)
# --broad-cutoff is distance between peaks to be merged
# -B probably makes a beggraph file for the outputs

macs3 callpeak -f BAMPE -t \
${BASE_PATH}/CTCF_10-20_shearing_R1.mLb.clN.sorted.bam \
${BASE_PATH}/CTCF_20-60_shearing_R1.mLb.clN.sorted.bam \
-c \
${BASE_PATH}/CTCF_10-20_shearing_input_R1.mLb.clN.sorted.bam \
${BASE_PATH}/CTCF_20-60_shearing_input_R1.mLb.clN.sorted.bam \
-g hs -n CTCF_chipseq_combined  -q 0.01 --outdir ../results/ctcf_wash_rnase
date
```


# We ran this in the bin directory with Run.sh file. 
# The data output from Macs was written to "results" folder


# (2) Create simple annotation file (.saf) for input into FeatureCounts
```{r loading in the consensus peak file (- RNase ) }

# loading in the narrow peak consensus peaks between two CTCF ChIPseq replicates (June 2022)
# Two inputs compared to two replicates and run on narrow peak instead of NF_CORE output broadPeak

ctcf_peaks <- import("results/CTCF_chipseq_combined_peaks.narrowPeak")

# The input to FeatureCounts is .saf (simple annotation file) which will quantify the reads in each peak
# we need a SAF file needs that requires these colnames:
# GeneID		Chr	Start	End	Strand
# creating SAF file below

ctcf_saf <- ctcf_peaks %>%
  as.data.frame() %>%
  # making sure to only grab normal chr not contigs (e.g., GL000009.2)
  filter(grepl("chr", seqnames)) %>%
  # renaming cols to fit format of .saf
  dplyr::rename(GeneID = name,
                Chr = seqnames,
                Start = start,
                End = end,
                Strand = strand) %>%
    dplyr::select(GeneID, Chr, Start, End, Strand)

# Writting out .saf
write.table(ctcf_saf, "results/ctcf_peaks_combined.saf", sep = "\t", quote = F, row.names = F)

```


# (3) Run feature counts on all bam files +/- rnase
# This will provide quntification of reads in each peak in consensus peaks
# For each sample -- these values can be used as input for deseq below
```{r feature counts across all bam files (+/- RNase)}

# ***** Once Run this can be loaded in to save time ***** #
# Start point
peak_counts <- read_rds("results/CTCF_feature_counts_output.rds")


# The ENCODE files are single-end reads and located in /data 
# create an object of all bam files (*** including inputs ***)
se_bams <- list.files("data/encode_bam", full.names = T)


# Running feature counts on the bam files from ENCODE that are single ended (-RNase)
# default is single ended but added isPairedEnd parameter to False

fc_se <- featureCounts(files = se_bams,
                    annot.ext = "results/ctcf_peaks_combined.saf",
                    isPairedEnd = FALSE,
                    nthreads = 6)


# We'll need to run featureCounts separately for the paired-end samples 
# Running our CTCF chip replicates and inputs through featureCounts 

# these files are far away so creating base-path 
base_path <- "/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/"

# Creating a list of all bams and file paths
pe_bams <- c(paste0(base_path, "CTCF_10-20_shearing_R1.mLb.clN.sorted.bam"), 
             paste0(base_path, "CTCF_20-60_shearing_R1.mLb.clN.sorted.bam"),
             paste0(base_path, "CTCF_10-20_shearing_input_R1.mLb.clN.sorted.bam"),
             paste0(base_path, "CTCF_20-60_shearing_input_R1.mLb.clN.sorted.bam"))

# Running feature counts on CTCF chip samples and inputs (June 2022)

fc_pe <- featureCounts(files = pe_bams,
                    annot.ext = "results/ctcf_peaks_combined.saf",
                    isPairedEnd = TRUE,
                    nthreads = 6)

# Now we need to merge the count tables for paired and single end reads
# Checking everything is in order
stopifnot(all(rownames(fc_pe$counts) == rownames(fc_se$counts)))

# merging data with Cbind to keep rows the same (peak name)
peak_counts <- cbind(fc_se$counts, fc_pe$counts)

# saving
write_rds(peak_counts, "results/CTCF_feature_counts_output.rds")

```

# Now that we have the counts, we can run DEseq to determine which peaks are different 
# between conditions (+/ - RNase)


# For now we're comparing CTCF ChIP-seq run in different labs, 
# We can for now test this approach with Encode being + RNase and our samples (June 2022 CTCF) as -RNase

# Note on normalization
# In the past, we've seen better results normalizing based on the total number of mapped reads
# An example of this is here:
https://github.com/msmallegan/on_rchip/blob/bf94a96b2579748c8f923efbf413a1c7eac00b94/analysis/01_quantification/01_quantification.Rmd#L271

# (4) Deseq to determine peaks that are differntial between conditions
```{r running deseq between our CTCF and encode CTCF two replicates each}

# creating a dataframe to be used in Deseq (DESeqDataSetFormMatrix)
sample_df <- data.frame(filename = colnames(peak_counts)) %>%
  # Cleanup colnames from peak_counts
  mutate(sample_name = filename,
         sample_name = gsub(".bam|.R1.mLb.clN.sorted|.shearing", "", sample_name),
         sample_name = gsub("CTCF.10.20", "rinn.CTCF.rep1", sample_name),
         sample_name = gsub("CTCF.20.60", "rinn.CTCF.rep2", sample_name)) %>%
  separate(sample_name, into = c("experiment", "antibody", "replicate", "condition")) %>%
  replace_na(list(condition = "IP"))

# Snyder vs Rinn differential test of IP samples 
# In the future this will be RNase+ vs RNase-
# The input is consensus peaks derived from RNase - condition
sample_df1 <- sample_df %>%
  filter(experiment %in% c("snyder", "rinn"),
         condition == "IP") %>%
  mutate(experiment = factor(experiment, levels = c("rinn", "snyder")))

# indexing peak counts from feature counts to snyder and rinn samples
peak_counts1 <- peak_counts[,sample_df1$filename]

# setting up experimental design for + or - RNase (currently snyder vs rinn : experiment col)
dds1 <- DESeqDataSetFromMatrix(countData = peak_counts1,
                              colData = sample_df1,
                              # diff on snyder IPs (2) vs rinn IPs (2)
                              design = ~ experiment)
# Running Deseq
dds1 <- DESeq(dds1)

# getting names
resultsNames(dds1)

# filter out intercept results
res1 <- results(dds1, name = "experiment_snyder_vs_rinn", tidy = TRUE)

# Bernstein vs Rinn differential test of IP samples 
# In the future this will be RNase+ vs RNase-
# The input is consensus peaks derived from RNase - condition

# creating sample dataframe for Deseq
sample_df2 <- sample_df %>%
  filter(experiment %in% c("bernstein", "rinn"),
         condition == "IP") %>%
  mutate(experiment = factor(experiment, levels = c("rinn", "bernstein")))

# indexing consensus peaks to bernstein and rinn IP samples
peak_counts2 <- peak_counts[,sample_df2$filename]

# setting up Deseq design with bernstein and rinn as model comparison for diff
dds2 <- DESeqDataSetFromMatrix(countData = peak_counts2,
                              colData = sample_df2,
                              design = ~ experiment)
# running deseq
dds2 <- DESeq(dds2)

# getting result names
resultsNames(dds2)

# filtering out intercept result
res2 <- results(dds2, name = "experiment_bernstein_vs_rinn", tidy = TRUE)

```

# (5) Calculating the log fold change over input for each comparison.
# We will input the consensus peaks for all +rnase and -rnase replicates to Deseq
# Deseq will then provide the "average" lfc of peaks over input across replicates.
```{r getting log fold change over input for consensus peaks in each condition}

# Rinn vs Input ( - RNase)
sample_rinn_input <- sample_df %>%
  filter(experiment %in% c("rinn")) %>%
  mutate(condition = factor(condition, levels = c("input", "IP")))

# indexing counts for the input and IP samples for one condition (- RNase)
peak_counts_rinn_input <- peak_counts[,sample_rinn_input$filename]

# setting up deseq to compare input to IP for -RNase
dds_rinn_input <- DESeqDataSetFromMatrix(countData = peak_counts_rinn_input,
                              colData = sample_rinn_input,
                              design = ~ condition)

# running deseq
dds_rinn_input <- DESeq(dds_rinn_input)

# getting result names
resultsNames(dds_rinn_input)

# exporting results "using tidy" to a nice data frame for plotting later 
res_rinn_input <- results(dds_rinn_input, name = "condition_IP_vs_input", tidy = TRUE)

# Snyder vs Input ( + Rnase)
sample_snyder_input <- sample_df %>%
  filter(experiment %in% c("snyder") |
         experiment == "rinn" & condition == "input") %>%
  mutate(condition = factor(condition, levels = c("input", "IP")))

# indexing into the snyder input and IP samples
peak_counts_snyder_input <- peak_counts[,sample_snyder_input$filename]

# setting up differential comparison of +RNase IPs and -rnase input 
#TODO use the same -rnase input even though the IP is +RNase 
dds_snyder_input <- DESeqDataSetFromMatrix(countData = peak_counts_snyder_input,
                              colData = sample_snyder_input,
                              design = ~ condition)

# running deseq across input and +rnase samples to calculate lfc
dds_snyder_input <- DESeq(dds_snyder_input)

# getting result names
resultsNames(dds_snyder_input)

# exporting the lfc and other values into handy data frame to plot later
res_snyder_input <- results(dds_snyder_input, name = "condition_IP_vs_input", tidy = TRUE)

# Bernstein vs Input as another replicate of the +rnase condition for plotting below
sample_bernstein_input <- sample_df %>%
  filter(experiment %in% c("bernstein") |
         experiment == "rinn" & condition == "input") %>%
  mutate(condition = factor(condition, levels = c("input", "IP")))

# indexing into Bernstein IP and input counts 
peak_counts_bernstein_input <- peak_counts[,sample_bernstein_input$filename]

# setting up deseq to get lfc of +RNase IP and input 
dds_bernstein_input <- DESeqDataSetFromMatrix(countData = peak_counts_bernstein_input,
                              colData = sample_bernstein_input,
                              design = ~ condition)

# running deseq
dds_bernstein_input <- DESeq(dds_bernstein_input)

# getting result names
resultsNames(dds_bernstein_input)

# exproting Bernstein (+rnase) lfc of consensus peaks over input for plotting below
res_bernstein_input <- results(dds_bernstein_input, name = "condition_IP_vs_input", tidy = TRUE)
```


# (6) Plotting -NRase LFC(IP/input) vs + RNase LFC(IP/input)
```{r}

# The first plot will be snyder vs rinn.
# In this case snyder is +Rnase and Rinn = -RNase
# also making a col if peak is sig between +/- RNase
snyder_vs_rinn_df <- res1 %>%
  dplyr::select(row, log2FoldChange, padj) %>%
  dplyr::rename(l2fc_snyder_vs_rinn = log2FoldChange,
                padj_snyder_vs_rinn = padj) %>%
  left_join(res_snyder_input %>%
              dplyr::select(row, log2FoldChange, padj) %>%
              dplyr::rename(l2fc_snyder_vs_input = log2FoldChange,
                            padj_snyder_vs_input = padj)) %>%
  left_join(res_rinn_input %>%
              dplyr::select(row, log2FoldChange, padj) %>%
              dplyr::rename(l2fc_rinn_vs_input = log2FoldChange,
                            padj_rinn_vs_input = padj))
snyder_vs_rinn_df <- snyder_vs_rinn_df %>%
  mutate(sig = padj_snyder_vs_rinn < 0.05)

# actually plotting
ggplot(snyder_vs_rinn_df, aes(x = l2fc_rinn_vs_input, y = l2fc_snyder_vs_input, color = sig)) +
  geom_point() +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  geom_abline(slope = 1)


# as a second comparison rinn (-rnase) and bernstein (+rnase)
bernstein_vs_rinn_df <- res1 %>%
  dplyr::select(row, log2FoldChange, padj) %>%
  dplyr::rename(l2fc_bernstein_vs_rinn = log2FoldChange,
                padj_bernstein_vs_rinn = padj) %>%
  left_join(res_bernstein_input %>%
              dplyr::select(row, log2FoldChange, padj) %>%
              dplyr::rename(l2fc_bernstein_vs_input = log2FoldChange,
                            padj_bernstein_vs_input = padj)) %>%
  left_join(res_rinn_input %>%
              dplyr::select(row, log2FoldChange, padj) %>%
              dplyr::rename(l2fc_rinn_vs_input = log2FoldChange,
                            padj_rinn_vs_input = padj))
bernstein_vs_rinn_df <- bernstein_vs_rinn_df %>%
  mutate(sig = padj_bernstein_vs_rinn < 0.05)

# Plotting
ggplot(bernstein_vs_rinn_df, aes(x = l2fc_rinn_vs_input, y = l2fc_bernstein_vs_input, color = sig)) +
  geom_point() +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  geom_abline(slope = 1)

```

# If we want to normalize the read counts from feature counts to total reads in the sample:
# we run samtools to count reads in each bam file.
# 260 flag means aligned reads
# C flag is counts

```{r}
# TODO figure out normalization and incorporate above.
```

```{bash}

#
# activating samtools in bash
module load samtools

#TODO Would be better to figure out the read flags in feature counts to get a normalization factor

# -c is counts
# -F aligned reads (primary)

samtools view -c -F 260 bernstein_CTCF_rep2.bam

# 13147668

samtools view -c -F 260 bernstein_CTCF_rep1.bam

# 12414498

```





