---
title: "CTCF Chipseq analysis"
author: "JR"
date: "6/26/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
library(ComplexHeatmap)
source("../../util/intersect_functions.R")
source()

```


# Finding overlaps between each of the CTCF datasets

```{r determining overlaps between each CTCF chip}
# loading in gencode annotations
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/data/genomes/gencode.v32.annotation.gtf")

# loading in CTCF new sample 1 peaks
CTCF_1 <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/CTCF_10-20_shearing_R1_peaks.broadPeak")

# how many peaks?
summary(CTCF_1)
# CTCF_1 has 29,365 peaks 

# determining the width of peaks
summary(width(CTCF_1))

# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 199.0   297.0   395.0   476.6   543.0 15889.0 

# loading in CTCF new sample 2 peaks
CTCF_2 <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/CTCF_20-60_shearing_R1_peaks.broadPeak")

# how many peaks
summary(CTCF_2)
# CTCF_2 has 34,446 peaks

# determining the width of peaks
summary(width(CTCF_2))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 170.0   256.0   361.0   449.7   546.0 12775.0
```
### Result: peak number and widths are pretty similar CTCF_2 (20-60) has more peaks [34,446 vs  29, 365]

```{r determining overlaps between each CTCF chip}
# finding overlaps between CTCF_1 & CTCF_2
ov <- findOverlaps(CTCF_1, CTCF_2)

# number of overlaps?
summary(ov)
# 26,511 overlapping peaks

# determining the overlaps of CTCF_1 in CTCF_2
table(table(ov@to))
# 26,372 of 29,365 peaks overlap from CTCF_1 to CTCF_2
# 68 peaks had 2 overlpas and 1 peak had 3 overlaps

percent_overlap_1 <- (26372 + (2*68) + 3) / 29365
# 90% overlap

# determining the overlaps of CTCF_2 in CTCF_1
table(table(ov@from))
# 26,215 of 34,446 peaks overlap from CTCF_2 to CTCF_1
# 146 peaks had 2 overlaps and 1 had 3

percent_overlap_2 <- (26215 + (2*146) + 3) / 34446
# 77% of peaks overlap in CTCF_2 in CTCF_1

# trying countoverlaps to compare
ov_count_overlaps <- countOverlaps(CTCF_1, CTCF_2)
table(ov_count_overlaps)

# similar answer


consensus_ctcf <- create_consensus_peaks(CTCF_1, CTCF_2)


```
### Result: there is good overlap betwen the new CTCF samples 90% for CTCF1 and 77% CTCF 2 





# Loading in ENCODE CTCF peaks (replicate 1 and 2 merged from Bernstein group)
```{r}

#TODO having trouble importing narrow peak files
encode_ctcf <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/encode_peak_files/ENCFF736NYC.bed.gz")

encode_peaks <- read.table("data/encode_peaks/ENCFF736NYC.bed.gz", sep = "\t")
stam_ctcf <- GRanges(seqnames = encode_peaks$V1,
                     ranges = IRanges(start = encode_peaks$V2,
                                      end = encode_peaks$V3))


# ctcf_peaks <- GenomicRanges::reduce(CTCF_1, CTCF_2)

table(countOverlaps(stam_ctcf, CTCF_1))
length(stam_ctcf)

stam_ov1 <- findOverlaps(CTCF_1, stam_ctcf)
length(unique(stam_ov1@from))/length(CTCF_1)

stam_ov2 <- findOverlaps(CTCF_2, stam_ctcf)
length(unique(stam_ov2@from))/length(CTCF_2)

# ENCODE 4 PIPELINE -- bed narrowPeak	conservative IDR thresholded peaks	1, 2
encode_peaks <- read.table("data/encode_peaks/ENCFF769AUF.bed.gz", sep = "\t")
bernstein_ctcf <- GRanges(seqnames = encode_peaks$V1,
                     ranges = IRanges(start = encode_peaks$V2,
                                      end = encode_peaks$V3))

bernstein_stam_ov1 <- findOverlaps(stam_ctcf, bernstein_ctcf)
length(unique(bernstein_stam_ov1@from))/length(stam_ctcf)

source("../../util/encode_functions.R")

stam_files_df <- encode_file_info("ENCSR000DWE", file_format = "bed") %>%
  filter(file_format_type == "narrowPeak",
   output_type == "IDR thresholded peaks",
   genome_assembly == "GRCh38",
   biological_replicates == 12) %>%
  mutate(full_download_url = paste0("https://www.encodeproject.org", download_url))
system(paste0("cd data/encode_peaks wget ", stam_files_df$full_download_url))
# https://www.encodeproject.org/files/ENCFF736NYC/@@download/ENCFF736NYC.bed.gz
```

# Metaplot of CTCF chip replicates

```{r}

peak_coverage_CTCF1 <- coverage(CTCF_1)


coverage_length <- elementNROWS(peak_coverage_CTCF1)

coverage_gr <- GRanges(seqnames = names(coverage_length),
                       IRanges(start = rep(1, length(coverage_length)), 
                               end = coverage_length))


all_promoters_gr <- rtracklayer::import("lncRNA_mrna_promoters.gtf")

all_promoters_gr <- subsetByOverlaps(all_promoters_gr, 
                                  coverage_gr, 
                                  type="within", 
                                  ignore.strand=TRUE)


chromosomes <- intersect(names(peak_coverage_CTCF1), unique(as.character(seqnames(all_promoters_gr))))

peak_coverage_CTCF1 <- peak_coverage_CTCF1[chromosomes]

all_promoters_ir <- as(all_promoters_gr, "IntegerRangesList")[chromosomes]


promoter_peak_view <- Views(peak_coverage_CTCF1, all_promoters_ir)

promoter_peak_view <- lapply(promoter_peak_view, function(x) t(viewApply(x, as.vector)))


promoter_peak_matrix <- do.call("rbind", promoter_peak_view)

dim(promoter_peak_matrix)
#TODO more peaks ??

minus_idx <- which(as.character(strand(all_promoters_gr)) == "-")

promoter_peak_matrix[minus_idx,] <- promoter_peak_matrix[minus_idx, ncol(promoter_peak_matrix):1]

promoter_peak_matrix <- promoter_peak_matrix[rowSums(promoter_peak_matrix) > 0,]

peak_sums <- colSums(promoter_peak_matrix)
# normalization of binding events to 1 across promoter window
peak_dens <- peak_sums/sum(peak_sums)
# Create a data frame in order to plot this. 
metaplot_df <- data.frame(x = -3e3:(3e3-1), dens = peak_dens)

ggplot(metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)



ggplot(metaplot_df, aes(x = x, y = dens)) + 
  geom_vline(xintercept = 0, lty = 2) + 
  geom_line(size = 1.5) + 
  ggtitle("CTCF_1 Promoter Metaplot") + 
  scale_x_continuous(breaks = c(-3000, 0, 3000),
                     labels = c("-3kb", "TSS", "+3kb"),
                     name = "") + 
  ylab("Peak frequency")




```


# Metaplot CTCF 2

```{r}


peak_coverage_CTCF2 <- coverage(CTCF_2)


coverage_length <- elementNROWS(peak_coverage_CTCF2)

coverage_gr <- GRanges(seqnames = names(coverage_length),
                       IRanges(start = rep(1, length(coverage_length)), 
                               end = coverage_length))


# all_promoters_gr <- rtracklayer::import("lncRNA_mrna_promoters.gtf")

all_promoters_gr <- subsetByOverlaps(all_promoters_gr, 
                                  coverage_gr, 
                                  type="within", 
                                  ignore.strand=TRUE)


chromosomes <- intersect(names(peak_coverage_CTCF2), unique(as.character(seqnames(all_promoters_gr))))

peak_coverage_CTCF2 <- peak_coverage_CTCF2[chromosomes]

all_promoters_ir <- as(all_promoters_gr, "IntegerRangesList")[chromosomes]


promoter_peak_view <- Views(peak_coverage_CTCF2, all_promoters_ir)

promoter_peak_view <- lapply(promoter_peak_view, function(x) t(viewApply(x, as.vector)))


promoter_peak_matrix <- do.call("rbind", promoter_peak_view)

dim(promoter_peak_matrix)
#TODO more peaks ??

minus_idx <- which(as.character(strand(all_promoters_gr)) == "-")

promoter_peak_matrix[minus_idx,] <- promoter_peak_matrix[minus_idx, ncol(promoter_peak_matrix):1]

promoter_peak_matrix <- promoter_peak_matrix[rowSums(promoter_peak_matrix) > 0,]

peak_sums <- colSums(promoter_peak_matrix)
# normalization of binding events to 1 across promoter window
peak_dens <- peak_sums/sum(peak_sums)
# Create a data frame in order to plot this. 
metaplot_df <- data.frame(x = -3e3:(3e3-1), dens = peak_dens)

ggplot(metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)



ggplot(metaplot_df, aes(x = x, y = dens)) + 
  geom_vline(xintercept = 0, lty = 2) + 
  geom_line(size = 1.5) + 
  ggtitle("CTCF_1 Promoter Metaplot") + 
  scale_x_continuous(breaks = c(-3000, 0, 3000),
                     labels = c("-3kb", "TSS", "+3kb"),
                     name = "") + 
  ylab("Peak frequency")




```

