---
title: "CTCF Chipseq analysis"
author: "JR"
date: "6/26/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
source("../../util/intersect_functions.R")

```


# Finding overlaps between each of the CTCF datasets

```{r determining overlaps between each CTCF chip}
# loading in gencode annotations
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/data/genomes/gencode.v32.annotation.gtf")

# loading in CTCF new sample 1 peaks
CTCF_1 <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/CTCF_10-20_shearing_R1_peaks.broadPeak")

# how many peaks?
summary(CTCF_1)
# CTCF_1 has 29,365 peaks 

# determining the width of peaks
summary(width(CTCF_1))

# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 199.0   297.0   395.0   476.6   543.0 15889.0 

# loading in CTCF new sample 2 peaks
CTCF_2 <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/CTCF_20-60_shearing_R1_peaks.broadPeak")

# how many peaks
summary(CTCF_2)
# CTCF_2 has 34,446 peaks

# determining the width of peaks
summary(width(CTCF_2))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 170.0   256.0   361.0   449.7   546.0 12775.0
```
### Result: peak number and widths are pretty similar CTCF_2 (20-60) has more peaks [34,446 vs  29, 365]

```{r determining overlaps between each CTCF chip}
# finding overlaps between CTCF_1 & CTCF_2
ov <- findOverlaps(CTCF_1, CTCF_2)

# number of overlaps?
summary(ov)
# 26,511 overlapping peaks

# determining the overlaps of CTCF_1 in CTCF_2
table(table(ov@to))
# 26,372 of 29,365 peaks overlap from CTCF_1 to CTCF_2
# 68 peaks had 2 overlpas and 1 peak had 3 overlaps

percent_overlap_1 <- (26372 + (2*68) + 3) / 29365
# 90% overlap

# determining the overlaps of CTCF_2 in CTCF_1
table(table(ov@from))
# 26,215 of 34,446 peaks overlap from CTCF_2 to CTCF_1
# 146 peaks had 2 overlaps and 1 had 3

percent_overlap_2 <- (26215 + (2*146) + 3) / 34446
# 77% of peaks overlap in CTCF_2 in CTCF_1

# trying countoverlaps to compare
ov_count_overlaps <- countOverlaps(CTCF_1, CTCF_2)
table(ov_count_overlaps)

# similar answer

```
### Result: there is good overlap betwen the new CTCF samples 90% for CTCF1 and 77% CTCF 2 





# Loading in ENCODE CTCF peaks (replicate 1 and 2 merged from Bernstein group)
```{r}

#TODO having trouble importing narrow peak files
encode_ctcf <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/ENCFF430KTH.Bed")

```



