---
title: "00_new_candidate_qc"
author: "JR"
date: "7/26/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(GenomicRanges)
library(tidyverse)
# library(Gviz)
library(IRanges)
source("../../util/intersect_functions.R")
source("../../util/plotting_functions.R")
```

First going to encode portal and grabbing latest "consensus peaks" 
between replicates for each candidate.

# Candidate 1: HDAC1
First download the consensus peak files of each candidate using wget.

```{bash}

# change to working dir
cd /scratch/Shares/rinn/JR/rchip/analysis/00_NEW_candidate_qc/data/HDAC1

# download 
HDAC1: wget https://www.encodeproject.org/files/ENCFF432KJA/@@download/ENCFF432KJA.bed.gz

# there are other we can add but look similar

# They are all here:
# https://www.encodeproject.org/search/?type=Experiment&control_type!=*&status=released&perturbed=false&target.label=HDAC1&biosample_ontology.term_name=K562&assay_title=TF+ChIP-seq&biosample_ontology.classification=cell%20line

# example:
# HDAC1_2: https://www.encodeproject.org/files/ENCFF661VOO/@@download/ENCFF661VOO.bed.gz

# let's unzip the peak file
gunzip ENCFF432KJA.bed.gz

```

Cool now we have a peak file. Let's read it and find good peaks for qPCR

```{r import peak file}

HDAC1_peaks <- read.table("data/HDAC1/ENCFF432KJA.bed")

HDAC1_peaks <- read_tsv("//scratch/Shares/rinn/JR/rchip/analysis/00_NEW_candidate_qc/data/HDAC1/ENCFF432KJA.bed", col_names = F)

# col values are :
# Chromosome, start, stop, name, score, strand, signalValue, pValue, qValue, peak_center

# adding colnames 

names(HDAC1_peaks) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')

# Note this is bed+4 format / .broadPeak files are .bed+3 -- no peak center
# the rest is the same

# Nice the data frame is reeadable and reachable 
# Now we want to find some good peaks in the raw data by eye
# We have tracks in UCSC or load in bigWig into IGV
# Mostly we want to find big, sig peaks that are in good areas of genome

```


# convert peak file to Granges for meta plot (or overlaps etc)
```{r}

HDAC1_1_gr <- GRanges(HDAC1_peaks$chromosome, IRanges(HDAC1_peaks$start, HDAC1_peaks$end))

test <- GRanges(seqnames = HDAC1_peaks$chromosome,
                ranges = IRanges(start=HDAC1_peaks$start,end=HDAC1_peaks$end))

```


check consistency between replicates
# replicate consistency
```{r replicate overlaps}

# TODO check manually by subsetByOverlaps of each peak file 
# TODO count # peaks and overlaps
# could add promoter overlap?

```


# metaplot of peaks
```{r}

# First we need promoters let's make them quick: (actually takes 5 min or so)
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2023/data/data/genomes/gencode.v32.annotation.gtf")

# create promoters 
promoters <- promoters(gencode_gr[gencode_gr$type == "gene"], 
                 upstream = 1000, downstream = 1000)

# making empty data frame for profile_tss to populate
HDAC1_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())

# running profile TSS 
HDAC1_metaplot_df <- profile_tss(HDAC1_1_gr , promoters)

#TODO if I add promoter_gr = promoters I get three of these ???

ggplot(HDAC1_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)


```


Trying RBP1 as comparison
```{r}

# loading in peak file

RBP1_peaks <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/RBP1_no_rnase_R1_peaks.broadPeak")

RBP1_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())

# running profile TSS 
RBP1_metaplot_df <- profile_tss(RBP1_peaks , promoters)

#TODO if I add promoter_gr = promoters I get three of these ???

ggplot(RBP1_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)


```



















# Valr metaplot stuff
```{r}
# Let's get the promoters in the valr format.
# We'll just get the TSS
tss <- promoters(gencode_gr[gencode_gr$type == "gene"], upstream = 0, downstream = 0) %>%
  gr_to_bed()
# We're just going to take the peak center.
HDAC1_peaks  <- read_bed("data/HDAC1/ENCFF432KJA.bed") %>%
  mutate(peak_center = start + X10)

hdac_tss_ov <- bed_intersect(HDAC1_peaks, tss) %>%
  mutate(meta_position = peak_center.x - start.y,
         meta_position = ifelse(strand.y == "-", meta_position * -1, meta_position))

hdac_tss_meta_peak_center <- hdac_tss_ov %>%
  group_by(meta_position) %>%
  summarize(npeaks = n())

ggplot(hdac_tss_meta_peak_center, aes(x = meta_position, y = npeaks)) +
  geom_point() +
  geom_smooth(span = 0.2)



```









# goal qPCR peaks to design primers for rchip.
the next steps are:

1) Sort on qValue -log10(FDR) -- thus bigger number more sig
2) Check in browser for peak quality etc... may take a few tries
-copy and paste peak coordinates and then look and see
3) note candidate regions to be tested by qPCR +/- RNase

# The browser is down?? Can't find them -- moving to meta plot




